### ISSUE TITLE
Package visibility changes in Android 11 break browser selection

### ISSUE THREAD

**Issue by M66B** [openid/AppAuth-Android#554]
Related Android issue:

https://issuetracker.google.com/issues/158914189

**Comment by hnljp** [comment]
I have documented it in https://github.com/openid/AppAuth-Android/pull/553
Add the following permission to your android manifest:
```xml
<uses-permission android:name="android.permission.QUERY_ALL_PACKAGES" />
```

**Comment by M66B** [comment]
@hnljp this will not be an options for apps published in the Play store because there will likely be new play store policies disallowing common apps with this permission.

https://developer.android.com/preview/privacy/package-visibility

**Comment by M66B** [comment]
A better workaround at this time is:

```
    <queries>
        <intent>
            <action android:name="android.intent.action.VIEW" />
            <data android:scheme="https" />
        </intent>
    </queries>
```

However, Google could disallow this for common apps (not launcher, not security related, etc) too.

**Comment by hnljp** [comment]
Why would you think that it would happen?
It will most likely "just" be added to the list of permissions that Android 11 users will need to accept that the app uses.

But yes, it would be nice with an option, where simply the default CustomTab capable browser would be chosen... But that would remove the whitelist/blacklist option.

**Comment by M66B** [comment]
Because it is documented here:

https://developer.android.com/preview/privacy/package-visibility#all-apps

**The following list gives some examples of use cases where the QUERY_ALL_PACKAGES permission is appropriate to include:**

* **Launcher apps**
* **Accessibility apps**
* **Browsers**
* **Peer-to-peer (P2P) sharing apps**
* **Device management apps**
* **Security apps**

**Comment by hnljp** [comment]
Yes, that is correct, and the queries solution at least makes sure that the app is only able to search for a limited set of apps.
Good point.

**Comment by M66B** [comment]
I am not sure that Google is going to allow whitelisting *android.intent.action.VIEW* though.

**Comment by hnljp** [comment]
I have changed it to:
```
    <queries>
        <intent>
            <action android:name="android.intent.action.VIEW" />
            <category android:name="android.intent.category.BROWSABLE" />
            <data android:scheme="https" />
        </intent>
    </queries>
```
This will make sure that it is only "fullbrowsers" that we get.
The results are already filtering, that the packages should support the category android.intent.category.BROWSABLE ... so by doing this, we will also get less packages that the AppAuth BrowserSelector needs to run though.

**Comment by M66B** [comment]
I have just tested this and it works fine.

Hopefully, Google is going to allow this kind of whitelisting.

**Comment by hnljp** [comment]
Yes, it should be better.

**Comment by fenrir-wangxin** [comment]
add <uses-permission android:name="android.permission.QUERY_ALL_PACKAGES" />

**Comment by braczkow** [comment]
Hi AppAuth team :)
Do you have any feedback from Google / users of your library if the _<queries>.._ workaround is a real risk for an App to be blocked by the PlayStore?

**Comment by agologan** [comment]
We have updated our code to include the correct `<queries>` block per the official documentation: https://developer.android.com/training/basics/intents/package-visibility-use-cases#check-browser-available in #613 
This will be auto-merged in your manifest.
